src/common/containers.h

```diff
@@ -0,0 +1,179 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*- 
+// vim: ts=8 sw=2 smarttab
+//
+// Ceph - scalable distributed file system
+//
+// Copyright (C) 2004-2006 Sage Weil <sage@newdream.net>
+//
+// This is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License version 2.1, as published by the Free Software
+// Foundation.  See file COPYING.
+//
+
+#ifndef CEPH_COMMON_TINY_VECTOR_H
+#define CEPH_COMMON_TINY_VECTOR_H
+
+#include <type_traits>
+
+#include "include/intarith.h"
+
+namespace ceph::containers {
+
+// tiny_vector - a CPU-friendly container like small_vector but for
+// mutexes, atomics and other non-movable things.
+//
+// The purpose of the container is store arbitrary number of objects
+// with absolutely minimal requirements regarding constructibility
+// and assignability while minimizing memory indirection.
+// There is no obligation for MoveConstructibility, CopyConstructibility,
+// MoveAssignability, CopyAssignability nor even DefaultConstructibility
+// which allows to handle std::mutexes, std::atomics or any type embedding
+// them.
+//
+// Few requirements translate into tiny interface. The container isn't
+// Copy- nor MoveConstructible. Although it does offer random access
+// iterator, insertion in the middle is not allowed. The maximum number
+// of elements must be known at run-time. This shouldn't be an issue in
+// the intended use case: sharding.
+//
+// Alternatives:
+//  1. std::vector<boost::optional<ValueT>> initialized with the known
+//     size and emplace_backed(). boost::optional inside provides
+//     the DefaultConstructibility. Imposes extra memory indirection.
+//  2. boost::container::small_vector + boost::optional always
+//     requires MoveConstructibility.
+//  3. boost::container::static_vector feed via emplace_back().
+//     Good for performance but enforces upper limit on elements count.
+//     For sharding this means we can't handle arbitrary number of
+//     shards (weird configs).
+//  4. std::unique_ptr<ValueT>: extra indirection together with memory
+//     fragmentation.
+
+template<typename Value,
+         std::size_t Capacity,
+         std::size_t Alignment = alignof(Value)>
+class tiny_vector {
+  using storage_unit_t = \
+    std::aligned_storage_t<sizeof(Value), Alignment>;
```

Comment:
We'll be going with [extended alignment](http://en.cppreference.com/w/cpp/language/object#Alignment) for e.g. `shard_vector`. AFAIK the standard makes it non-obligatory for compiler vendors, albeit this feature should be available on most platforms. In any case, this deserves a comment.

Reply:
FIXED.